test/distributed/test_multi_threaded_pg.py

```diff
@@ -27,6 +27,36 @@ def test_broadcast_object_list(self):
         dist.broadcast_object_list(object_list=object_list)
         self.assertEqual(99, object_list[0])
 
+    def test_collective_error_on_rank_zero(self):
+        @spawn_threads_and_init_comms(world_size=4)
+        def _test_method(self):
+            input_tensor = torch.ones(3, 3) * dist.get_rank()  # perform 1st all gather
+            output_tensors = [torch.empty_like(input_tensor) for _ in range(dist.get_world_size())]
+            dist.all_gather(output_tensors, input_tensor)
+
+            if dist.get_rank() == 0:
+                raise AssertionError("Mimic real test failure.")  # fail on rank 0
+
+            dist.all_gather(output_tensors, input_tensor)  # perform 2nd all gather
+
+        with self.assertRaisesRegex(AssertionError, "Mimic real test failure."):
+            _test_method(self)
+
+    def test_collective_error_on_rank_non_zero(self):
+        @spawn_threads_and_init_comms(world_size=4)
+        def _test_method(self):
+            input_tensor = torch.ones(3, 3) * dist.get_rank()  # perform 1st all gather
+            output_tensors = [torch.empty_like(input_tensor) for _ in range(dist.get_world_size())]
+            dist.all_gather(output_tensors, input_tensor)
+
+            if dist.get_rank() == 1:
```

Comment:
can we raise all ranks other than rank 0?

Reply:
is there a need to have two tests about this? 1 non-zero rank and 1 zero rank failure seems identical if you do that?

Reply:
since all collective operation happens on rank 0, non-zero rank failure and zero rank failure had different issues before this fix.